From 610b1801c1fe7d7504562ff8528062c26eeda2c3 Mon Sep 17 00:00:00 2001
From: Giovanni Campagna <gcampagna@src.gnome.org>
Date: Sat, 12 Jan 2013 23:49:05 +0100
Subject: [PATCH] Fix compilation with srcdir != builddir

We must look for marshal and enum templates in the source directory,
and we must not add the $(srcdir) prefix to sources that live in
the build directory.
---
 src/Makefile.am | 37 +++++++++++++++++++++++--------------
 1 file changed, 23 insertions(+), 14 deletions(-)

diff --git a/src/Makefile.am b/src/Makefile.am
index e4aeb63..79ed708 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -8,22 +8,28 @@
 # autogenerated files
 grl-marshal.h: grl-marshal.list
 	$(AM_V_GEN) $(GLIB_GENMARSHAL) --prefix grl_marshal	\
-	--header grl-marshal.list > $@
+	--header $< > $@
 
-grl-marshal.c: grl-marshal.h grl-marshal.list
+grl-marshal.c: grl-marshal.list grl-marshal.h
 	@echo "#include \"grl-marshal.h\"" > $@
 	$(AM_V_GEN) $(GLIB_GENMARSHAL) --prefix grl_marshal	\
-	--body grl-marshal.list >> $@
+	--body $< >> $@
 
 enum_headers = grl-source.h grl-caps.h grl-operation-options.h data/grl-media.h
 
 grl-type-builtins.h: $(enum_headers) grl-type-builtins.h.template
-	$(AM_V_GEN) $(GLIB_MKENUMS) --template grl-type-builtins.h.template	\
-	$(enum_headers) > $@
+	$(AM_V_GEN) (cd $(srcdir); $(GLIB_MKENUMS) --template grl-type-builtins.h.template	\
+	$(enum_headers) ) > $@
 
 grl-type-builtins.c: grl-type-builtins.h $(enum_headers) grl-type-builtins.c.template
-	$(AM_V_GEN) $(GLIB_MKENUMS) --template grl-type-builtins.c.template	\
-	$(enum_headers) > $@
+	$(AM_V_GEN) (cd $(srcdir); $(GLIB_MKENUMS) --template grl-type-builtins.c.template	\
+	$(enum_headers) ) > $@
+
+BUILT_SOURCES = \
+	grl-type-builtins.h \
+	grl-type-builtins.c \
+	grl-marshal.h	    \
+	grl-marshal.c
 
 lib_LTLIBRARIES = lib@GRL_NAME@.la
 
@@ -41,12 +47,14 @@ lib@GRL_NAME@_la_LDFLAGS =				\
 	-version-info $(GRL_LT_VERSION)	\
 	-no-undefined
 
-lib@GRL_NAME@_la_SOURCES =					\
+lib@GRL_NAME@_la_SOURCES = \
+	$(shipped_sources) \
+	$(BUILT_SOURCES)
+
+shipped_sources = 						\
 	grl-plugin.c grl-plugin-priv.h		\
 	grl-registry.c grl-registry-priv.h	\
 	grl-metadata-key.c grl-metadata-key-priv.h		\
-	grl-type-builtins.c grl-type-builtins.h			\
-	grl-marshal.c grl-marshal.h				\
 	grl-operation.c grl-operation.h				\
 	grl-operation-priv.h grl-sync.c						\
 	grl-source.c			\
@@ -68,7 +76,7 @@ data_c_sources =		\
 	data/grl-media-box.c	\
 	data/grl-config.c
 
-lib@GRL_NAME@_la_SOURCES += $(data_c_sources)
+shipped_sources += $(data_c_sources)
 
 lib@GRL_NAME@incdir =	\
 	$(includedir)/@GRL_NAME@
@@ -134,14 +142,15 @@ INTROSPECTION_COMPILER_ARGS = --includedir=$(srcdir) --includedir=$(srcdir)/data
 # introspection support
 if HAVE_INTROSPECTION
 introspection_sources = \
-	$(lib@GRL_NAME@inc_HEADERS)	\
-	$(lib@GRL_NAME@_la_SOURCES)
+	$(addprefix $(srcdir)/,$(lib@GRL_NAME@inc_HEADERS))	\
+	$(addprefix $(srcdir)/,$(shipped_sources))	\
+	$(BUILT_SOURCES)
 
 Grl-@GRL_MAJORMINOR@.gir: lib@GRL_NAME@.la
 Grl_@GRL_MAJORMINOR_NORM@_gir_INCLUDES = GObject-2.0 GModule-2.0
 Grl_@GRL_MAJORMINOR_NORM@_gir_CFLAGS = $(lib@GRL_NAME@_la_CFLAGS)
 Grl_@GRL_MAJORMINOR_NORM@_gir_LIBS = lib@GRL_NAME@.la
-Grl_@GRL_MAJORMINOR_NORM@_gir_FILES = $(addprefix $(srcdir)/,$(introspection_sources))
+Grl_@GRL_MAJORMINOR_NORM@_gir_FILES = $(introspection_sources)
 INTROSPECTION_GIRS += Grl-@GRL_MAJORMINOR@.gir
 
 girdir = @INTROSPECTION_GIRDIR@
-- 
1.8.0.1

